### ISSUE TITLE
zombie wrexecd processes for killed and/or exited jobs

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#1431]
Something is causing apparently dead wreck daemons to stick around, they're hammering a core each, and I think maybe hammering the kvs on the flux instance as well.  Pulling some data now.

**Comment by trws** [comment]
They're stuck in an infinite loop waiting for `ioplex:complete()` to return true, which apparently it never does.  Each time it fails, it goes into reactor once, then tries again.

**Comment by dongahn** [comment]
We saw a similar issue in #1400. You are using a patched version so this is a different leftover issue?

**Comment by grondo** [comment]
Ugh. Sorry about that. Waiting to pull all io from kvs to send to output
file, but possibly an error condition has the code confused...

**Comment by trws** [comment]
It’s related, but not the same.  In this case some of the wrexecd 
processes are running even for jobs that have completed or been killed 
and successfully marked as such.

**Comment by trws** [comment]
@garlick and I were just digging into this, it looks like the reactor is 
hitting a fatal condition, but the return code from the reactor:once is 
never checked, so it keeps looping on it.  The major tipoff was that 30% 
of the runtime was in strerror.

Do you know a good way to get the lua function names for backtraces by 
the way @grondo? I was trying to figure out what the reactor is actually 
up to in there, but all of the traces other than generic handler 
functions are sans-symbol, which I’m guessing means they’re lua 
functions?

**Comment by trws** [comment]
Note here, whatever the reactor is doing seemed to pin the master broker on KVS interactions.  In fact this seems to be the root cause of the slowdown issues with long running flux instances for splash.

Broker 0 perf output on kvs transactions:
```
-   48.22%     0.00%  flux-broker-0  kvs.so                 [.] kvstxn_link_dirent                                                                                                           ▒
   - kvstxn_link_dirent                                                                                                                                                                      ▒
      - 33.88% cache_entry_get_treeobj                                                                                                                                                       ▒
         - treeobj_decodeb                                                                                                                                                                   ▒
            - 27.44% json_loadb                                                                                                                                                              ▒
                 parse_json                                                                                                                                                                  ▒
               - parse_value                                                                                                                                                                 ▒
                  - 27.26% parse_value                                                                                                                                                       ▒
                     + 22.00% parse_value                                                                                                                                                    ▒
                     + 2.06% lex_scan                                                                                                                                                        ▒
                     + 2.02% json_object_set_new_nocheck                                                                                                                                     ▒
            + 6.34% treeobj_validate                                                                                                                                                         ▒
      + 8.01% treeobj_deep_copy                                                                                                                                                              ▒
      + 6.33% treeobj_insert_entry                                                                                                                                                           ▒
-   36.19%     0.01%  flux-broker-0  kvs.so                 [.] kvstxn_unroll                                                                                                                ▒
   - 36.18% kvstxn_unroll                                                                                                                                                                    ▒
      - 36.18% kvstxn_unroll                                                                                                                                                                 ▒
         - 36.18% kvstxn_unroll                                                                                                                                                              ▒
            - 36.16% kvstxn_unroll                                                                                                                                                           ▒
               - 36.15% kvstxn_unroll                                                                                                                                                        ▒
                  - 36.12% kvstxn_unroll                                                                                                                                                     ▒
                     + 18.92% store_cache.isra.0                                                                                                                                             ▒
                     + 9.59% kvstxn_unroll                                                                                                                                                   ▒
                     + 7.55% json_object_iter_set_new                                                                                                                                        ▒
```

wrexecd perf output spinning:

```
-  100.00%     0.00%  wrexecd  liblua-5.1.so          [.] lua_pcall                                                                                                                          ▒
     lua_pcall                                                                                                                                                                               ▒
     0xfbc8                                                                                                                                                                                  ▒
     0xe784                                                                                                                                                                                  ▒
     _init                                                                                                                                                                                   ▒
   - 0xf95c                                                                                                                                                                                  ▒
      - 67.97% 0x1ebcc                                                                                                                                                                       ▒
         - 65.35% 0xf478                                                                                                                                                                     ▒
            - 63.24% l_flux_reactor_start                                                                                                                                                    ▒
               - 32.72% flux_reactor_run                                                                                                                                                     ▒
                  - 29.86% ev_run                                                                                                                                                            ▒
                     - 26.23% ev_invoke_pending                                                                                                                                              ▒
                        - 20.89% check_cb                                                                                                                                                    ▒
                           - 18.44% handle_cb                                                                                                                                                ▒
                              - 17.31% handle_cb                                                                                                                                             ▒
                                 + 17.09% strerror                                                                                                                                           ▒
                           + 1.87% get_pollevents                                                                                                                                            ▒
                        + 3.05% prepare_cb                                                                                                                                                   ▒
                       0.83% epoll_poll                                                                                                                                                      ▒
                  + 1.16% __kernel_clock_gettime                                                                                                                                             ▒
               + 23.75% l_pushresult                                                                                                                                                         ▒
               + 2.68% luaL_checkudata                                                                                                                                                       ▒
               + 1.75% flux_get_reactor                                                                                                                                                      ▒
               + 1.17% flux_aux_get                                                                                                                                                          ▒
           0.57% 0xf20c                                                                                                                                                                      ▒
      - 20.79% 0x1e6b0                                                                                                                                                                       ▒
         - 16.84% 0x1c5bc                                                                                                                                                                    ▒
            - 16.41% 0xf8f8                                                                                                                                                                  ▒
               - 14.84% 0xf478                                                                                                                                                               ▒
                  + 10.33% l_wreck_index                                                                                                                                                     ▒
                  + 1.93% l_flux_index                                                                                                                                                       ▒
                    1.24% __strcmp_power9                                                                                                                                                    ▒
           0.91% 0x1a92c                                                                                                                                                                     ▒
           0.71% 0x1ce40                                                                                                                                                                     ▒
      - 2.86% 0x1e6f4                                                                                                                                                                        ▒
           0.81% 0x1a92c                                                                                                                                                                     ▒
      + 2.00% 0x1df40                                                                                                                                                                        ▒
        1.15% 0x1da78                                                                                                                                                                        ▒
        0.57% 0x1da70                                                                                                                                                                        ▒


```

full perf data for both, and a core file for the wreck daemon:
[perf-data.zip](https://github.com/flux-framework/flux-core/files/1881638/perf-data.zip)
I have the broker core file for the running process, but it is 14GB, so I can't really attach it.

**Comment by grondo** [comment]
On Thu, Apr 5, 2018, 12:48 PM Tom Scogland <notifications@github.com> wrote:

>
> Do you know a good way to get the lua function names for backtraces by
> the way @grondo? I was trying to figure out what the reactor is actually
> up to in there, but all of the traces other than generic handler
> functions are sans-symbol, which I’m guessing means they’re lua
> functions
>


Not off the top of my head, sorry!
Usually the C side has a good hint to where I haven't had to look at actual
Lua debuggers before.

**Comment by grondo** [comment]
This was probably fixed by 50b7be8228b25febe0b7b099b084e820ef027265, though it is possible that the zombie wrexecd were encountering another problem. Ok to close and reopen a new issue when we see this again?

**Comment by trws** [comment]
Works for me, that’s definitely where they were all hanging when I saw 
it happen before so if it comes back it’s probably a different issue.
